internal/controllers/machineset/machineset_controller.go

```diff
@@ -352,6 +358,131 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 	}
 
+	// Adjust the managedFields of the Machines to make them compatible with SSA.
+	for idx := range filteredMachines {
+		m := filteredMachines[idx]
+		if !m.DeletionTimestamp.IsZero() {
+			continue
+		}
+		if err := r.adjustManagedFields(ctx, m); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to adjust the managedFields of the Machine %q", m.Name)
+		}
+	}
+
+	// Update the in-place propagated fields on existing machines.
+	for idx := range filteredMachines {
+		m := filteredMachines[idx]
+		if !m.DeletionTimestamp.IsZero() {
+			continue
+		}
+		log := log.WithValues("Machine", klog.KObj(m))
+		updatedMachine := r.computeDesiredMachine(machineSet, m)
+		patchOptions := []client.PatchOption{
+			client.ForceOwnership,
+			client.FieldOwner(machineSetManagerName),
+		}
+		if err := r.Client.Patch(ctx, updatedMachine, client.Apply, patchOptions...); err != nil {
+			log.Error(err, "Failed to update existing Machine")
+			return ctrl.Result{}, errors.Wrap(err, "failed to update Machine")
+		}
+		filteredMachines[idx] = updatedMachine
+	}
+
+	// Update the InfraMachine objects of the machines.
+	for idx := range filteredMachines {
+		m := filteredMachines[idx]
+		if !m.DeletionTimestamp.IsZero() {
+			continue
+		}
+		infraMachine, err := external.Get(ctx, r.Client, &m.Spec.InfrastructureRef, m.Namespace)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to get infra machine object %q for machine %q", klog.KRef(m.Namespace, m.Spec.InfrastructureRef.Name), klog.KObj(m))
+		}
+		// Adjust the managed fields to make the infra machine object SSA compatible.
+		if err := r.adjustManagedFields(ctx, infraMachine); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to update managedfields of infra machine %q for machine %q", klog.KRef(m.Namespace, m.Spec.InfrastructureRef.Name), klog.KObj(m))
+		}
+		updatedInfraMachine, err := external.ComputeObjectFromTemplate(ctx, &external.ComputeObjectFromTemplateInput{
+			Client:      r.Client,
+			TemplateRef: &machineSet.Spec.Template.Spec.InfrastructureRef,
+			Namespace:   machineSet.Namespace,
+			ClusterName: machineSet.Spec.ClusterName,
+			OwnerRef: &metav1.OwnerReference{
+				APIVersion: clusterv1.GroupVersion.String(),
+				Kind:       "MachineSet",
+				Name:       machineSet.Name,
+				UID:        machineSet.UID,
+			},
+			Labels:      machineLabels(machineSet),
+			Annotations: machineSet.Spec.Template.Annotations,
+		})
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to compute updated infra machine %q for machine %q", klog.KRef(m.Namespace, m.Spec.InfrastructureRef.Name), klog.KObj(m))
+		}
+		// Since the InfraMachine already exists, use that name.
+		// Set the UID to the existing infra machine object to ensure that we only update and avoid accidentally
+		// creating an infra machine.
+		// TODO: The resourceVersion of the updatedInfraMachine will be empty. Is that okay?
```

Comment:
Good question. We also never did set it in the topology reconciler.

I think given that we always want to write our intent independent of what the object on the server-side looks like it should be fine.